// ku4node-kernel/src/geometry/vector.js
var $ = require("../base/asserters"),
    $str = require("../base/str"),
    $math = require("../base/math"),
    $class = require("../base/class")
    $coord = require("./coord");

function vector(x, y) {
    if (!$.isNumber(x) || !$.isNumber(y))
        throw new Error($str.format("at $.vector({0},{1})", x, y));

    vector.base.call(this, x, y);

    this._lengthSquared = calculateLengthSquared(this, x, y);
    this._length = calculateLength(this, this._lengthSquared);
    this._unitNormalX = calculateUnitNormal(this, x);
    this._unitNormalY = calculateUnitNormal(this, y);
}

vector.prototype = {
    magnatude: function(){ return this.get("length"); },
    equals: function(other) {
        return (other instanceof vector) &&
            ((this._x === other.x()) && (this._y === other.y()));
    },
    normal: function() { return new vector(this._unitNormalX, this._unitNormalY); },
    invert: function() { return new vector(this.x() * -1, this.y() * -1); },
    norm: function() { return new vector(Math.abs(this.x()), Math.abs(this.y())); },
    perpendicular: function(){ return new vector(this.y() * -1, this.x()); },
    isZero: function() { return this.x() == 0 && this.y() == 0; },
    add: function(other) { return new vector(this.x() + other.x(), this.y() + other.y()); },
    dot: function(other) { return (this.x() * other.x()) + (this.y() * other.y()); },
    perpendicularAtTo: function(other) {
        var p = other.add(this.projectionOfOnto(other).invert());
        return new vector(p.x(), p.y());
    },
    projectionOfOnto: function(other) {
        var p = other.normal().scale(this.dot(other.normal()));
        return new vector(p.x(), p.y());
    },
    scale: function(scalar) {
        return new vector((this.x() * scalar), (this.y() * scalar));
    },
    unitNormalDot: function(other) {
        return (this.normal().x() * other.normal().x()) +
            (this.normal().y() * other.normal().y());
    },
    reflect: function(incident){
        if(incident.isZero()) return this;
        var inorm = incident.normal()
        return this.add(inorm.scale(2*(inorm.dot(this))).invert());
    },
    round: function(decimal){
        var d = decimal || 0;
        return new vector($math.round(this.x(), d), $math.round(this.y(), d));
    }
}
$class.extend(vector, $coord.Class);

function calculateLength (v, lengthSquared) {
    if (v.isZero()) return 0;
    return Math.sqrt(lengthSquared);
}
function calculateLengthSquared(v, x, y) {
    if (v.isZero()) return 0;
    return Math.pow(x, 2) + Math.pow(y, 2)
}
function calculateUnitNormal (v, scalar) {
    if (v.isZero()) return 0;
    return scalar / v.magnatude();
}

module.exports = function(x, y) { return new vector(x, y); }
module.exports.Class = vector;
module.exports.zero = function() { return new vector(0,0); }
module.exports.random = function(seedx, seedy){
    var x = seedx * Math.random(), y = seedy * Math.random();
    return new vector(x, y);
}
