// ku4node-kernel/lib/base/asserters.js
module.exports = {
    isArray: function(x) {
        return x instanceof Array;
    },
    isBool: function(x) {
        return (/boolean/i.test(typeof (x)));
    },
    isDate: function(x) {
        return x instanceof Date;
    },
    isNumber: function(x) {
        return (/number/i.test(typeof (x))) && !isNaN(x);
    },
    isObject: function(x) {
        return  this.exists(x) &&
                (/object/i.test(typeof (x)));
    },
    isFunction: function(x) {
        return (x instanceof Function);
    },
    isString: function(x) {
        return (/string/i.test(typeof (x))) ||
                x instanceof String;
    },
    isZero: function(n) {
        return n === 0;
    },
    isEven: function(n) {
        return (this.isNullOrEmpty(n) ||
                this.isDate(n) ||
                isNaN(n) ||
                this.isZero(n))
                    ? false
                    : n % 2 === 0;
    },
    isOdd: function(n) {
        return (this.isNullOrEmpty(n) ||
                this.isDate(n) ||
                isNaN(n) ||
                this.isZero(n))
                    ? false
                    : !this.isEven(n);
    },
    isNull: function(x) {
        return x === null;
    },
    isUndefined: function(x) {
        return (/undefined/i.test(typeof (x)));
    },
    isEmpty: function(s) {
        return  this.isString(s) &&
                this.isZero(s.split(/\B/).length);
    },
    isNullOrEmpty: function(s) {
        return !this.exists(s) ||
                this.isEmpty(s);
    },
    exists: function(x) {
        return  (x !== null) &&
                (!this.isUndefined(x));
    },
    xor: function(a, b) {
        return !a != !b;
    }
}
