// ku4node-kernel/src/finance/money.js
var $ = require("../base/asserters"),
    $str = require("../base/str"),
    $math = require("../base/math");

function money(amt, type) {
    if (isNaN(amt)) throw new $.exception("arg", $str.format("$.money requires a number. Passed {0}", amt));
    var x = amt.toString().split(/\./), d = x[0], c = x[1];
    function cents(c) { return (amt < 0) ? -c : c; }

    this._cents = ($.exists(c)) ? cents(parseFloat("." + c)) : 0;
    this._dollars = parseInt(d);
    this._type = type || "$";
    this._value = amt;
}
money.prototype = {
    cents: function(){ return this._cents; },
    dollars: function(){ return this._dollars; },
    type: function(){ return this._type; },
    value: function(){ return this._value; },
    add: function(other) {
        checkType(this, other);
        return new money(this._value + other.value());
    },
    divide: function(value) {
        if(!$.isNumber(value))
            throw new Error();
        return new money(this._value / value);
    },
    equals: function(other) {
        return (this.isOfType(other)) && (this._value == other.value());
    },
    isOfType: function(other) {
        return this._type == other.type();
    },
    isGreaterThan: function(other) {
        checkType(this, other);
        return this._value > other.value();
    },
    isLessThan: function(other) {
        checkType(this, other);
        return this._value < other.value();
    },
    multiply: function(value) {
        if(!$.isNumber(value))
            throw new Error();
        return new money(this._value * value);
    },
    round: function() {
        return new money($math.round(this.value, -2));
    },
    roundDown: function() {
        return new money($math.roundDown(this.value, -2));
    },
    roundUp: function() {
        return new money($math.roundUp(this.value, -2));
    },
    subtract: function(other) {
        checkType(this, other);
        return new money(this._value - other.value());
    },
    toString: function() {
        var format = (this.value < 0) ? "({0}{1}.{2})" : "{0}{1}.{2}";
        return $str.format(format, this._type, formatDollars(this), formatCents(this));
    }
}

function checkType(m, other) {
    if (!m.isOfType(other)) throw new Error("Invalid operation on non-conforming currencies.");
}
function formatDollars(m) {
    var dollars = m.dollars(),
        anount = (m.cents() >= .995) ? (dollars + 1) : dollars,
        s = anount.toString(),
        d = s.replace(/\-/, "").split(/\B/).reverse(),
        l = d.length,
        b = l > 3,
        i = 0,
        a = [];
    while (i < l) {
        a[a.length] = d[i]; i++;
        if (!$.exists(d[i])) break;
        if ((i % 3 == 0) && b) a[a.length] = ",";
    }
    return $str.build.apply(this, a.reverse());
}
function formatCents(m) {
    var C = $math.round(m.cents(), -3),
        s = C.toString(),
        c = s.replace(/\-|(0\.)/g, "").concat("0").split(/\B/), l = c.length;
    if ($.isZero(l) || C >= .995) return "00";
    if (l < 2) return "0" + c[0];
    return (parseInt(c[2]) > 4) ? c[0] + (parseInt(c[1]) + 1) : c[0] + c[1];
}

function canParse(v){
    try {
        $.money.parse(v);
        return true;
    }
    catch(e){ return false; }
}
function parse(str) {
    if($.isNumber(str)) return new money(str);
    var b = /(\(.*\))|(\-)/.test(str),
        i = (b) ? 1 : 0,
        u = str.match(/[^\d\.\,\-]/g) || [],
        U = $.exists(u[i]) ? u[i] : "$",
        n = parseFloat(str.replace(/[^\d\.]/g, "")),
        v = (b) ? -n : n;
    return new money(v, U);
}

module.exports = function(number, type){ return new money(number, type); }
module.exports.zero = function() { return new money(0); }
module.exports.isMoney = function(o) { return o instanceof money; }
module.exports.canParse =canParse;
module.exports.parse = parse;
module.exports.tryParse = function(o){
    return canParse(o) ? parse(o) : null;
}
